### PR TITLE
EL-593: Allow different people to have housing costs and housing benefits

### PR DESCRIPTION
**Pull request by patrick-laa**
## What

[Link to story](https://dsdmoj.atlassian.net/browse/EL-593)

Allow negative housing costs when applicant has partner (so that applicant can have housing benefits that offset partner's housing costs)

### CODE REVIEW

File: app/services/collators/housing_costs_collator.rb [hunk 1]
```diff
@@ -1,9 +1,9 @@
 module Collators
   class HousingCostsCollator
     class << self
-      def call(disposable_income_summary:, gross_income_summary:, submission_date:, person:)
+      def call(disposable_income_summary:, gross_income_summary:, submission_date:, person:, allow_negative_net: false)
```
Comment (starswan) [inline app/services/collators/housing_costs_collator.rb:4]: I think this default can be removed as all callers provide it

File: app/services/collators/housing_costs_collator.rb [hunk 2]
```diff
@@ -1,9 +1,9 @@
...
         housing_calculator = Calculators::HousingCostsCalculator.new(disposable_income_summary:, gross_income_summary:,
-                                                                     submission_date:, person:)
+                                                                     submission_date:, person:, allow_negative_net:)
```
Comment (starswan) [inline app/services/collators/housing_costs_collator.rb:6]: As this calculator is only called in one place, the flag could be avoided at this level by moving the [0, net_housing_costs] logic up one level (to this class) and making the decision here. WDYT?

File: app/services/calculators/housing_costs_calculator.rb [hunk 3]
```diff
@@ -5,11 +5,12 @@ class HousingCostsCalculator
 
     delegate :housing_cost_outgoings, to: :@disposable_income_summary
 
-    def initialize(disposable_income_summary:, gross_income_summary:, submission_date:, person:)
+    def initialize(disposable_income_summary:, gross_income_summary:, submission_date:, person:, allow_negative_net: false)
```
Comment (starswan) [inline app/services/calculators/housing_costs_calculator.rb:8]: This doesn't need defaulting - it is supplied by all callers (except tests, which can be fixed)

File: app/services/collators/housing_costs_collator.rb (hunk 2 above)
Comment (patrick-laa) [reply app/services/collators/housing_costs_collator.rb]: Sounds good!

**Review by starswan — APPROVED** [review]

**Pull request merged**